components/style/atomic_refcell.rs

```diff
@@ -2,121 +2,358 @@
-/// allocates an OS-specific primitive and delegates operations to it. However,
-/// parking_lot provides a pure-rust implementation of the standard
-/// synchronization primitives, with uncontended borrows compiling down to a
-/// single atomic operation. This is exactly how we would build an atomic
-/// RefCell, so we newtype it with some API sugar.
-pub struct AtomicRefCell<T>(RwLock<T>);
-
-pub struct AtomicRef<'a, T: 'a>(RwLockReadGuard<'a, T>);
-unsafe impl<'a, T> StableAddress for AtomicRef<'a, T> {}
-
-impl<'a, T> Deref for AtomicRef<'a, T> {
-    type Target = T;
-    fn deref(&self) -> &T {
-        self.0.deref()
-    }
+/// A threadsafe analogue to RefCell.
+pub struct AtomicRefCell<T: ?Sized> {
+    borrow: AtomicUsize,
+    value: UnsafeCell<T>,
 }
 
-pub struct AtomicRefMut<'a, T: 'a>(RwLockWriteGuard<'a, T>);
-unsafe impl<'a, T> StableAddress for AtomicRefMut<'a, T> {}
+impl<T> AtomicRefCell<T> {
+    /// Creates a new `AtomicRefCell` containing `value`.
+    #[inline]
+    pub fn new(value: T) -> AtomicRefCell<T> {
+        AtomicRefCell {
+            borrow: AtomicUsize::new(0),
+            value: UnsafeCell::new(value),
+        }
+    }
 
-impl<'a, T> Deref for AtomicRefMut<'a, T> {
-    type Target = T;
-    fn deref(&self) -> &T {
-        self.0.deref()
+    /// Consumes the `AtomicRefCell`, returning the wrapped value.
+    #[inline]
+    pub fn into_inner(self) -> T {
+        debug_assert!(self.borrow.load(atomic::Ordering::Acquire) == 0);
+        unsafe { self.value.into_inner() }
     }
 }
 
-impl<'a, T> DerefMut for AtomicRefMut<'a, T> {
-    fn deref_mut(&mut self) -> &mut T {
-        self.0.deref_mut()
+impl<T: ?Sized> AtomicRefCell<T> {
+    /// Immutably borrows the wrapped value.
+    #[inline]
+    pub fn borrow(&self) -> AtomicRef<T> {
+        AtomicRef {
+                value: unsafe { &*self.value.get() },
+                borrow: AtomicBorrowRef::new(&self.borrow),
+        }
+    }
+
+    /// Mutably borrows the wrapped value.
+    #[inline]
+    pub fn borrow_mut(&self) -> AtomicRefMut<T> {
+        AtomicRefMut {
+                value: unsafe { &mut *self.value.get() },
+                borrow: AtomicBorrowRefMut::new(&self.borrow),
+        }
+    }
+
+    /// Returns a raw pointer to the underlying data in this cell.
+    #[inline]
+    pub fn as_ptr(&self) -> *mut T {
+        self.value.get()
     }
 }
 
-impl<'a, T: 'a + Debug> Debug for AtomicRef<'a, T> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{:?}", self.0.deref())
+//
+// Core synchronization logic. Keep this section small and easy to audit.
+//
+
+const HIGH_BIT: usize = !(::std::usize::MAX >> 1);
+const MAX_FAILED_BORROWS: usize = HIGH_BIT + (HIGH_BIT >> 1);
+
+struct AtomicBorrowRef<'b> {
+    borrow: &'b AtomicUsize,
+}
+
+impl<'b> AtomicBorrowRef<'b> {
+    #[inline]
+    fn new(borrow: &'b AtomicUsize) -> Self {
+        let new = borrow.fetch_add(1, atomic::Ordering::AcqRel) + 1;
```

Comment:
Yes, `Acquire` is sufficient here. You only need to prevent memory operations after this line from being reordered before acquiring the guard.